In [1]:
!cd

C:\Users\arvin\LLM_From_Scratch_solo\tokenization


In [15]:
with open("the-verdict.txt","r") as f:
    text_R = f.read()

In [16]:
print("Total number of character:", len(text_R))
print(text_R[:99])

Total number of character: 20479
I HAD always thought Jack Gisburn rather a cheap genius--though a good fellow enough--so it was no 


In [11]:
import re
text = "Hello, world. This, is a test."
result = re.split(r'(\s)', text)
print(result)
result = [item for item in result if item.strip() != ""]
print(result)

['Hello,', ' ', 'world.', ' ', 'This,', ' ', 'is', ' ', 'a', ' ', 'test.']
['Hello,', 'world.', 'This,', 'is', 'a', 'test.']


In [17]:
preprocessed = re.split(r'([,.:;?_!"()\']|--|\s)', text_R)
preprocessed = [item.strip() for item in preprocessed if item.strip()]
print(preprocessed)

['I', 'HAD', 'always', 'thought', 'Jack', 'Gisburn', 'rather', 'a', 'cheap', 'genius', '--', 'though', 'a', 'good', 'fellow', 'enough', '--', 'so', 'it', 'was', 'no', 'great', 'surprise', 'to', 'me', 'to', 'hear', 'that', ',', 'in', 'the', 'height', 'of', 'his', 'glory', ',', 'he', 'had', 'dropped', 'his', 'painting', ',', 'married', 'a', 'rich', 'widow', ',', 'and', 'established', 'himself', 'in', 'a', 'villa', 'on', 'the', 'Riviera', '.', '(', 'Though', 'I', 'rather', 'thought', 'it', 'would', 'have', 'been', 'Rome', 'or', 'Florence', '.', ')', '"', 'The', 'height', 'of', 'his', 'glory', '"', '--', 'that', 'was', 'what', 'the', 'women', 'called', 'it', '.', 'I', 'can', 'hear', 'Mrs', '.', 'Gideon', 'Thwing', '--', 'his', 'last', 'Chicago', 'sitter', '--', 'deploring', 'his', 'unaccountable', 'abdication', '.', '"', 'Of', 'course', 'it', "'", 's', 'going', 'to', 'send', 'the', 'value', 'of', 'my', 'picture', "'", 'way', 'up', ';', 'but', 'I', 'don', "'", 't', 'think', 'of', 'that', ',

In [18]:
all_words = sorted(set(preprocessed))
vocab_size = len(all_words)
print(vocab_size)

1130


In [21]:
vocab = {token:integer for integer,token in enumerate(all_words)}
for i, item in enumerate(vocab.items()):
    print(item)
    if i >= 50:
        break


('!', 0)
('"', 1)
("'", 2)
('(', 3)
(')', 4)
(',', 5)
('--', 6)
('.', 7)
(':', 8)
(';', 9)
('?', 10)
('A', 11)
('Ah', 12)
('Among', 13)
('And', 14)
('Are', 15)
('Arrt', 16)
('As', 17)
('At', 18)
('Be', 19)
('Begin', 20)
('Burlington', 21)
('But', 22)
('By', 23)
('Carlo', 24)
('Chicago', 25)
('Claude', 26)
('Come', 27)
('Croft', 28)
('Destroyed', 29)
('Devonshire', 30)
('Don', 31)
('Dubarry', 32)
('Emperors', 33)
('Florence', 34)
('For', 35)
('Gallery', 36)
('Gideon', 37)
('Gisburn', 38)
('Gisburns', 39)
('Grafton', 40)
('Greek', 41)
('Grindle', 42)
('Grindles', 43)
('HAD', 44)
('Had', 45)
('Hang', 46)
('Has', 47)
('He', 48)
('Her', 49)
('Hermia', 50)


In [2]:
from importlib.metadata import version
import tiktoken
print("tiktoken version:", version("tiktoken"))

tiktoken version: 0.12.0


In [4]:
tokenizer = tiktoken.get_encoding("gpt2")

text = (
    "Hello, do you like tea? <|endoftext|> In the sunlit terraces"
    "of someunknownPlace."
)
integers = tokenizer.encode(text, allowed_special={"<|endoftext|>"})
print(integers)

[15496, 11, 466, 345, 588, 8887, 30, 220, 50256, 554, 262, 4252, 18250, 8812, 2114, 1659, 617, 34680, 27271, 13]


In [5]:
print(len(text))
print(len(integers))

80
20


In [6]:
strings = tokenizer.decode(integers)
print(strings)

Hello, do you like tea? <|endoftext|> In the sunlit terracesof someunknownPlace.


BPE tokenizer or byte-pair encoding tokenizer is based upon a sophisticated tokenization scheme which was used to train llms such as ChatGPT 

It utilizes the <| endoftext |> token to join two different segments of text 

the BPE tokenizer used to train GPT-2,3 and the original model in ChatGPT had a vocabulary size of 50,257 with <| endoftext |> being assigned the largest token id of 50,256

It utilizes an algorithm which breaks down the unknown words not in its vocabulary into a sequence of smaller sub-words or even individual characters to encode it.

Example:


In [7]:
integers = tokenizer.encode("Kabjngvf :;wfwg")
print(integers)
strings = tokenizer.decode(integers)
print(strings)

[42, 397, 73, 782, 85, 69, 1058, 26, 86, 44482, 70]
Kabjngvf :;wfwg


As we see the "unknown" string was broken down by the tokenizer into a sequence of smaller segments which were effectively decoded too later on

In [6]:
import tiktoken
tokenizer = tiktoken.get_encoding("gpt2")
with open("the-verdict.txt", "r", encoding="utf-8") as f:
    raw_text = f.read()
enc_text = tokenizer.encode(raw_text)
print(len(enc_text))

5145


In [7]:
enc_sample = enc_text[50:]

context_size = 5      
x = enc_sample[:context_size]
y = enc_sample[1:context_size+1]
print(f"x: {x}")
print(f"y:      {y}")

x: [290, 4920, 2241, 287, 257]
y:      [4920, 2241, 287, 257, 4489]


In [8]:
for i in range(1, context_size+1):
    context = enc_sample[:i]
    desired = enc_sample[i]
    print(context, "---->", desired)

[290] ----> 4920
[290, 4920] ----> 2241
[290, 4920, 2241] ----> 287
[290, 4920, 2241, 287] ----> 257
[290, 4920, 2241, 287, 257] ----> 4489


In [9]:
for i in range(1, context_size+1):
    context = enc_sample[:i]
    desired = enc_sample[i]
    print(tokenizer.decode(context), "---->", tokenizer.decode([desired]))

 and ---->  established
 and established ---->  himself
 and established himself ---->  in
 and established himself in ---->  a
 and established himself in a ---->  vill


In [2]:
from torch.utils.data import Dataset, DataLoader

In [1]:
import tiktoken
with open("the-verdict.txt", "r", encoding="utf-8") as f:
    raw_text = f.read()
tokenizer = tiktoken.get_encoding("gpt2")

In [4]:
import torch

In [ ]:
class GPTDataset(Dataset):
    def __init__(self, text, tokenizer, context_size, stride):
        self.input_ids = []
        self.target_ids = []
        tokenized_text = tokenizer.encode(text)
        for i in range(0, len(tokenized_text) - context_size, stride):
            input_id = tokenized_text[i:i+context_size]
            target_id = tokenized_text[i+1:i+context_size+1]
            self.input_ids.append(input_id)
            self.target_ids.append(target_id)

    def __len__(self):
        return len(self.input_ids)

    def __getitem__(self, sample_id):
        return self.input_ids[sample_id], self.target_ids[sample_id]

In [ ]:
dataset = GPTDataset(
    raw_text, tokenizer,
    context_size=4,
    stride=1
)

In [7]:
dataset[0]

([40, 367, 2885, 1464], [367, 2885, 1464, 1807])

In [8]:
len(dataset)

5141

In [9]:
raw_text[:100]

'I HAD always thought Jack Gisburn rather a cheap genius--though a good fellow enough--so it was no g'

In [13]:
tokenizer.decode(dataset[0][0])
tokenizer.decode(dataset[0][1])

' HAD always thought'

In [15]:
def data_loader(text, batch_size, context_window = 256, stride = 64, shuffle = False,num_workers = 3):
    tokenizer = tiktoken.get_encoding("gpt2")
    dataset = GPTDataset(
        text, tokenizer,
        context_size=context_window,
        stride=stride
    )
    return DataLoader(dataset, batch_size=batch_size, shuffle=shuffle, num_workers=num_workers)
    

In [22]:
dataloader = data_loader(raw_text, batch_size=4, context_window = 256, stride = 64, shuffle=True, num_workers=14)

In [24]:
dataloader.dataset[0]

([40,
  367,
  2885,
  1464,
  1807,
  3619,
  402,
  271,
  10899,
  2138,
  257,
  7026,
  15632,
  438,
  2016,
  257,
  922,
  5891,
  1576,
  438,
  568,
  340,
  373,
  645,
  1049,
  5975,
  284,
  502,
  284,
  3285,
  326,
  11,
  287,
  262,
  6001,
  286,
  465,
  13476,
  11,
  339,
  550,
  5710,
  465,
  12036,
  11,
  6405,
  257,
  5527,
  27075,
  11,
  290,
  4920,
  2241,
  287,
  257,
  4489,
  64,
  319,
  262,
  34686,
  41976,
  13,
  357,
  10915,
  314,
  2138,
  1807,
  340,
  561,
  423,
  587,
  10598,
  393,
  28537,
  2014,
  198,
  198,
  1,
  464,
  6001,
  286,
  465,
  13476,
  1,
  438,
  5562,
  373,
  644,
  262,
  1466,
  1444,
  340,
  13,
  314,
  460,
  3285,
  9074,
  13,
  46606,
  536,
  5469,
  438,
  14363,
  938,
  4842,
  1650,
  353,
  438,
  2934,
  489,
  3255,
  465,
  48422,
  540,
  450,
  67,
  3299,
  13,
  366,
  5189,
  1781,
  340,
  338,
  1016,
  284,
  3758,
  262,
  1988,
  286,
  616,
  4286,
  705,
  1014,
  510,
  26,
  